src/dsp/interface.ts

```diff
@@ -1,56 +1,40 @@
 
 export class DataPrep {}
 
-export class DataSourcePlugin {
-  static getLibDSPName() {
-    // SHOULD align with zkgraph-lib/dsp/ethereum/index.ts
-    // return ['zkmain_name_in_lib', 'asmain_name_in_lib']
-    throw new Error('default: getLibDSPName not implemented in DSP.')
-  }
-
-  static async prepareData(zkgraphYaml: ZkGraphYaml, prepareParams: Record<string, any>): Promise<any> {
-    throw new Error('default: prepareData not implemented in DSP.')
-  }
-
-  static fillExecInput(input: any, zkgraphYaml: any, dataPrep: any) {
-    throw new Error('default: fillInput not implemented in DSP.')
-  }
-
-  static fillProveInput(input: any, zkgraphYaml: any, dataPrep: any) {
-    throw new Error('default: fillInput not implemented in DSP.')
-  }
-
-  static toProveDataPrep(execDataPrep: any, execResult: any) {
-    throw new Error('default: toProveDataPrep not implemented in DSP.')
-  }
+export abstract class DataSourcePlugin<EP extends object, PP extends object, PRP extends object, DPP extends object> {
+  abstract getLibDSPName(): string
+  abstract prepareData(zkgraphYaml: ZkGraphYaml, prepareParams: PRP): Promise<DPP>
+  abstract fillExecInput(input: Input, zkgraphYaml: ZkGraphYaml, dataPrep: DPP): Input
+  abstract fillProveInput(input: Input, zkgraphYaml: ZkGraphYaml, dataPrep: DPP): Input
+  abstract toProveDataPrep(execDataPrep: DPP, execResult: any): DPP
 
-  static toPrepareParams(...args: any[]) {
-    throw new Error('default: toPrepareParams not implemented in DSP.')
+  abstract execParams: KeyofToArray<EP>
+  toExecParams(params: Record<string, any>) {
+    return dspParamsNormalize(this.execParams as string[], params) as EP
   }
 
-  static execParams: string[] = []
-
-  static toExecParams(params: Record<string, any>) {
-    throw new Error('default: toExecParams not implemented in DSP.')
+  abstract proveParams: KeyofToArray<PP>
+  toProveParams(params: Record<string, any>) {
+    return dspParamsNormalize(this.proveParams as string[], params) as PP
   }
 
-  static proveParams: string[] = []
-
-  static toProveParams(params: Record<string, any>) {
-    throw new Error('default: toProveParams not implemented in DSP.')
+  toPrepareParamsFromExecParams(execParams: EP): Promise<PRP> {
+    return this.toPrepareParams(execParams, 'exec')
   }
 
-  static async toPrepareParamsFromExecParams(execParams: any): Promise<any> {
-    throw new Error('default: toPrepareParamsFromExecParams not implemented in DSP.')
+  toPrepareParamsFromProveParams(proveParams: PP): Promise<PRP> {
+    return this.toPrepareParams(proveParams, 'prove')
   }
 
-  static async toPrepareParamsFromProveParams(proveParams: any): Promise<any> {
-    throw new Error('default: toPrepareParamsFromProveParams not implemented in DSP.')
-  }
+  abstract toPrepareParams(params: EP, type: 'exec'): Promise<PRP>
+  abstract toPrepareParams(params: PP, type: 'prove'): Promise<PRP>
+  abstract toPrepareParams(params: EP | PP, type: 'exec' | 'prove'): Promise<PRP>
```

Comment:
can we only keep the combined type version, rm L37-38?

Reply:
function overloading can have a better development experience.
such as pass `type` is "exec", params type must be `EP`.

Reply:
is it? i think passing params `PP`, type = "exec", can also pass the check isn't it?

Reply:
If that's the case, should remove L39 (the combined type version), only keep 37-38.